### ISSUE TITLE
Arrow functions, needs a fix.

### ISSUE THREAD

**Issue by hemanth** [google/traceur-compiler#1664]
```js
new (() => {})
```

According to the spec it must throw : `TypeError: () => {} is not a constructor` but traceur is [not](https://google.github.io/traceur-compiler/demo/repl.html#new%20%28%28%29%20%3D%3E%20{}%29%3B%0A%0A)

**Comment by johnjbarton** [comment]
Please post a reference to the specification, thanks.

**Comment by UltCombo** [comment]
@johnjbarton 

Relevant arrow function spec:

- https://people.mozilla.org/~jorendorff/es6-draft.html#sec-arrow-function-definitions-runtime-semantics-evaluation
    > 4\. Let closure be FunctionCreate(Arrow, parameters, ConciseBody, scope, strict).

- https://people.mozilla.org/~jorendorff/es6-draft.html#sec-functioncreate
    > 2\. If kind is not Normal, let allocKind be "non-constructor".
    > 3\. Else let allocKind be "normal".
    > 4\. Let F be FunctionAllocate(functionPrototype, Strict, allocKind).

- https://people.mozilla.org/~jorendorff/es6-draft.html#sec-functionallocate
    > 4\. If functionKind is "non-constructor", then
    > &nbsp;&nbsp;&nbsp; a. Let functionKind be "normal".
    > &nbsp;&nbsp;&nbsp; b. Let needsConstruct be false.
    > [...]
    > 10\. If needsConstruct is true, then
    > &nbsp;&nbsp;&nbsp; a. Set F’s [[Construct]] internal method to the definition specified in 9.2.3.
 
The above condition is never met, so arrow functions do not have a [[Construct]] internal method set.

Relevant spec parts about the `new` operator:

- https://people.mozilla.org/~jorendorff/es6-draft.html#sec-evaluatenew
    > 9\. If IsConstructor (constructor) is false, throw a TypeError exception.

- https://people.mozilla.org/~jorendorff/es6-draft.html#sec-isconstructor
    > 3\. If argument has a [[Construct]] internal method, return true.
    > 4\. Return false.

**Comment by arv** [comment]
Not only arrow functions, accessors and concise methods should also throw when [[Construct]]ed.

However, this is not doable with ES5 semantics.

**Comment by johnjbarton** [comment]
Why can't we reject static forms? I can see that `new expr` where `expr = (() => {})` can be hard, but the static case in the original post seems doable. (Not a big priority to be sure).

**Comment by arv** [comment]
Static checks are doable but they will just catch the most naive cases. If we really want to enforce this we should rewrite `new expr(args)` to `$traceurRuntime.construct(expr, args)`.

**Comment by domenic** [comment]
> If we really want to enforce this we should rewrite new expr(args) to $traceurRuntime.construct(expr, args).

That is probably a good idea.

Alternately you could go for the half-assed instanceof check (i.e. each arrow function gets a unique name and then inserted `if (this instanceof myName) { throw new TypeError; }`)

**Comment by arv** [comment]
@domenic Yeah that is also an option. But I can come up with contrived cases where that would throw when it shouldn't.

 All of these options are bad for performance so maybe a strict/slow mode could be provided.

**Comment by hemanth** [comment]
Thanks for positing the reference links @UltCombo.

`if (this instanceof myName) { throw new TypeError; }` :+1:
